Gems/Atom/RHI/WebGPU/Code/Source/RHI/ShaderResourceGroupPool.cpp

```diff
@@ -5,30 +5,137 @@
+            group.m_compiledData.push_back(bindGroup);
+        }
+
+        return result;
+    }
+
+    RHI::ResultCode ShaderResourceGroupPool::CompileGroupInternal(
+        RHI::DeviceShaderResourceGroup& groupBase, const RHI::DeviceShaderResourceGroupData& groupData)
+    {
+        auto& group = static_cast<ShaderResourceGroup&>(groupBase);
+
+        group.UpdateCompiledDataIndex(m_currentIteration);
+        BindGroup& bindGroup = *group.m_compiledData[group.GetCompileDataIndex()];
+
+        const RHI::ShaderResourceGroupLayout* layout = groupData.GetLayout();
+
+        auto const& shaderBufferList = layout->GetShaderInputListForBuffers();
+        for (uint32_t groupIndex = 0; groupIndex < static_cast<uint32_t>(shaderBufferList.size()); ++groupIndex)
         {
-            return aznew ShaderResourceGroupPool();
+            const RHI::ShaderInputBufferIndex index(groupIndex);
+            auto bufViews = groupData.GetBufferViewArray(index);
+            uint32_t binding = shaderBufferList[groupIndex].m_registerId;
+            bindGroup.UpdateBufferViews(binding, bufViews);
         }
 
-        void ShaderResourceGroupPool::OnFrameEnd()
+        auto const& shaderImageList = layout->GetShaderInputListForImages();
+        for (uint32_t groupIndex = 0; groupIndex < static_cast<uint32_t>(shaderImageList.size()); ++groupIndex)
         {
-            Base::OnFrameEnd();
+            const RHI::ShaderInputImageIndex index(groupIndex);
+            auto imgViews = groupData.GetImageViewArray(index);
+            uint32_t binding = shaderImageList[groupIndex].m_registerId;
+            bindGroup.UpdateImageViews(binding, imgViews);
+        }        
+
+        auto const& shaderSamplerList = layout->GetShaderInputListForSamplers();
+        for (uint32_t groupIndex = 0; groupIndex < static_cast<uint32_t>(shaderSamplerList.size()); ++groupIndex)
+        {
+            const RHI::ShaderInputSamplerIndex index(groupIndex);
+            auto samplerArray = groupData.GetSamplerArray(index);
+            uint32_t binding = shaderSamplerList[groupIndex].m_registerId;
+            bindGroup.UpdateSamplers(binding, samplerArray);
         }
 
-        void ShaderResourceGroupPool::ShutdownInternal()
+        // WebGPU doesn't support static sampler, so we just use normal samplers
+        auto const& shaderStaticSamplerList = layout->GetStaticSamplers();
+        for (uint32_t groupIndex = 0; groupIndex < static_cast<uint32_t>(shaderStaticSamplerList.size()); ++groupIndex)
         {
-            Base::ShutdownInternal();
+            uint32_t binding = shaderStaticSamplerList[groupIndex].m_registerId;
+            bindGroup.UpdateSamplers(binding, AZStd::span<const RHI::SamplerState>(&shaderStaticSamplerList[groupIndex].m_samplerState, 1));
         }
 
-        void ShaderResourceGroupPool::ShutdownResourceInternal(RHI::DeviceResource& resourceBase)
+        auto constantData = groupData.GetConstantData();
```

Comment:
I get that auto is allowed in this codebase for non-iterator, but in my opinion it does hurt the comprehension in this file given that there are many of them and you cannot infer the types behind easily